lib/src/facts/linux/processor_resolver.cc

```diff
@@ -13,65 +14,155 @@ using namespace std;
 using namespace boost::filesystem;
 
 namespace lth_file = leatherman::file_util;
+namespace lth_util = leatherman::util;
 
 namespace facter { namespace facts { namespace linux {
+    static bool split_line(string const& line, string& key, string& value) {
+        // Split the line on colon
+        auto pos = line.find(":");
+        if (pos == string::npos) {
+            return false;
+        }
+        key = line.substr(0, pos);
+        boost::trim(key);
+        value = line.substr(pos + 1);
+        boost::trim(value);
+      
+        return true;
+    }
 
-    processor_resolver::data processor_resolver::collect_data(collection& facts)
+    processor_resolver::ArchitectureType processor_resolver::architecture_type(data const& data, std::string const& root)
     {
-        auto result = posix::processor_resolver::collect_data(facts);
+        if (!data.isa.empty()) {
+            return (boost::starts_with(data.isa, "ppc64")) ? POWER : X86;
+        }
 
-        unordered_set<string> cpus;
-        lth_file::each_subdirectory("/sys/devices/system/cpu", [&](string const& cpu_directory) {
-            ++result.logical_count;
-            string id = lth_file::read((path(cpu_directory) / "/topology/physical_package_id").string());
-            boost::trim(id);
-            if (id.empty() || cpus.emplace(move(id)).second) {
-                // Haven't seen this processor before
-                ++result.physical_count;
+        // use /proc/cpuinfo
+        unordered_set<string> to_be_seen;
+        bool seen_all = false;
+        lth_file::each_line(root + "/proc/cpuinfo", [&](string& line) {
+            // if we already know that we're on a Power machine, we can just skip
+            // the remaining lines
+            if (seen_all) {
```

Comment:
I think you could `return false` here to skip the remaining lines without continuing to loop over them